Deep Search Agent Refactor: SearchPipeline Integration & Schema Standardization

muthukumaranR opened the pull request:
# Deep Search Agent Refactor: SearchPipeline Integration & Schema Standardization

## Overview

This PR refactors the `DeepLitSearchAgent` to use the standardized `SearchPipeline` architecture, removes direct search tool dependencies, and standardizes I/O schemas for better consistency across the AKD framework.

##  Key Changes

### 0. Changes Cherrypicked + Resolved Merge Conflicts with Develop from a branch that had unreviewed code `bugfix/deep-search`

e6e1f9942e9c8ca9de352019010266010c265ded
476023725bc95316642f2a6c5c632682cac10d62
29405ab05c6c7a5029166f8f58b86ff66847e6e9
9eccca5d588939c2ec622e5d2304a570eabda3c7
55d5965a1959e69f16c31712e33fb0751702a05b
b2ff0b41b334c7dd5c09acfc7100b489058a69a0
8107a8fd9117560cfda8c64c73a3d1b18e74f735
1bdb079680c80b6b4fbef87744412881b351cdc5
e45b847db0a0e05446a7f1cf5e6380c350bd40b8
8a415290982ccd7380dbe51d8a157e3fd5ff04c3
85cc0d9e9aa4b8176555810a435a8297917ee171
200fd2bf604e12835ece6e5d90ff2960440a9ff9
4d7e0eacabcd2bc1c90542e4dd4cb3aedb7e8a97


### 1. Search Tool Abstraction & Pipeline Integration

**Removed direct search tool dependencies:**
- `SearxNGSearchTool`
- `SemanticScholarSearchTool`
- `LinkRelevancyAssessor`
- `SimplePDFScraper` & `SimpleWebScraper`

**Replaced with:**
- `SearchPipeline` - unified search interface that handles all search operations
- Dependency injection support for custom search pipelines
- Cleaner separation of concerns

### 2. Schema Standardization

**Updated `LitSearchAgentOutputSchema`:**
```python
class LitSearchAgentOutputSchema(SearchAgentOutputSchema):
    extra: Dict[str, Any] = Field(
        default_factory=dict,
        description="Extra metadata and synthesis information",
    )
```

**Removed custom schema within DeepSearchAgent:**
- `DeepLitSearchAgentOutputSchema` (custom schema)
- Uses standardized `LitSearchAgentOutputSchema` with `extra` field

**`SearchResultItem` usage:**
- Consistent search result format across all search agents
- Standardized metadata handling through `extra` fields

### 3. Research Synthesis Integration

**Enhanced output format:**
```python
return LitSearchAgentOutputSchema(
    results=results_as_dicts,
    category=params.category,
    iterations_performed=iterations,
    extra={
        "research_report": synthesis.research_report,
        "key_findings": synthesis.key_findings,
        "sources_consulted": synthesis.sources_consulted,
        "evidence_quality_score": synthesis.evidence_quality_score,
        "citations": synthesis.citations,
    }
)
```

### 4. Component Architecture Improvements

**Refactored embedded components:**
- `TriageComponent` - Enhanced debug logging and readability
- `ClarificationComponent` - Improved schema and field handling
- `InstructionBuilderComponent` - Added clarifications support
- `ResearchSynthesisComponent` - Updated token limits and performance
- `ContentCondensationComponent` - Enhanced functionality

### 5. Test Improvements

**Updated test suite:**
- Removed search tool-specific dependency tests
- Added SearchPipeline integration tests
- Fixed schema validation issues in mapping tests
- Optimized integration test performance (reduced iterations, timeouts)

**Key test fixes:**
```python
# Before: Used ExtractionDTO objects
lit_output = LitSearchAgentOutputSchema(
    results=[ExtractionDTO(source="...", result={...})],
)

# After: Uses dictionary format with required category field
lit_output = LitSearchAgentOutputSchema(
    results=[{"source": "...", "content": "..."}],
    category="science",
)
```

## 📁 Files Changed

### Core Agent Changes
- `akd/agents/search/deep_search.py` - Main refactor (390 lines removed, cleaner code)
- `akd/agents/search/_base.py` - Added `extra` field to output schema

### Component Updates
- `akd/agents/search/components/` - All components enhanced with better logging and functionality

### Configuration & Prompts
- `akd/configs/prompts.py` - Refactored for clarity and improved structure

### Documentation & Examples
- `examples/deep_search_test.py` - **New** comprehensive example
- `notebooks/deep_search_testing.ipynb` - Enhanced with execution outputs
- `docs/issn_whitelist_full.json` - Added complete ISSN whitelist

### Test Updates
- `tests/agents/search/test_deep_search.py` - Reduced from complex tool testing to clean pipeline testing
- `tests/mapping/test_mappers.py` - Fixed schema validation issues

## Usage Example

### Basic Usage
```python
from akd.agents.search.deep_search import DeepLitSearchAgent, DeepLitSearchAgentConfig
from akd.agents.search._base import LitSearchAgentInputSchema

# Create configuration
config = DeepLitSearchAgentConfig(
    max_research_iterations=2,
    quality_threshold=0.6,
    auto_clarify=False,
)

# Initialize agent (uses SearchPipeline automatically)
agent = DeepLitSearchAgent(config=config)

# Run research
input_params = LitSearchAgentInputSchema(
    query="machine learning applications in drug discovery",
    max_results=5
)

result = await agent.arun(input_params)

# Access results
print(f"Found {len(result.results)} results")
print(f"Research report: {result.extra['research_report']}")
print(f"Key findings: {result.extra['key_findings']}")
```

### Usage with Custom Pipeline
```python
from akd.tools.search.pipeline import SearchPipeline

# Use custom search pipeline
custom_pipeline = SearchPipeline(config=my_config)
agent = DeepLitSearchAgent(
    config=config,
    search_pipeline=custom_pipeline
)
```

## Example Script 

Run the example script:
```bash
python examples/deep_search_test.py
```

Run the test suite:
```bash
# Unit tests
pytest tests/agents/search/test_deep_search.py::TestDeepLitSearchAgent -v

# Integration tests (requires API keys)
pytest tests/agents/search/test_deep_search.py::TestDeepLitSearchAgentRealLLM -v

# Mapping tests
pytest tests/mapping/test_mappers.py::TestRealAgentMappings::test_full_agent_pipeline_mapping -v
```
---

muthukumaranR committed:
Refactor DeepLitSearchAgent to enhance scraping and source validation capabilities

- Introduced asyncio for improved asynchronous operations.
- Updated import statements to include new scrapers and resolvers.
- Added optional source validation feature based on ISSN.
- Enhanced the scraping logic to utilize a composite scraper for better content fetching.
- Implemented error handling for source validation initialization.
- Improved debug logging for better traceability of query and content fetching processes.

muthukumaranR committed:
Remove unused import of ContentCondenser from __init__.py in search components

muthukumaranR committed:
Refactor ClarifyingAgentInputSchema and ClarifyingAgentOutputSchema for improved readability

- Reformatted field definitions for better alignment and clarity.
- Enhanced debug logging in ClarificationComponent to provide more detailed output previews.
- Updated comments for consistency and clarity.

muthukumaranR committed:
Update ContentCondensationInputSchema and ContentCondensationComponent for enhanced functionality

- Increased the maximum token limit in ContentCondensationInputSchema from 4000 to 40000 to allow for larger condensed outputs.
- Added debug logging in ContentCondensationComponent to preview input and output content, improving traceability during the condensation process.

muthukumaranR committed:
Add clarifications field and enhance debug logging in InstructionBuilderComponent

- Introduced a new optional field `clarifications` in InstructionBuilderInputSchema to capture user feedback during instruction building.
- Enhanced debug logging in InstructionBuilderComponent to preview both input and output, improving traceability and debugging capabilities.

muthukumaranR committed:
Refactor ResearchSynthesisComponent and update token limit for improved performance

- Removed the `sources_consulted` field from `ResearchSynthesisOutputSchema` to streamline the output structure.
- Increased the `max_tokens` limit in `ResearchSynthesisAgentConfig` from 4000 to 60000 to accommodate larger input sizes.
- Enhanced debug logging in `ResearchSynthesisComponent` to provide previews of input and output, improving traceability during synthesis operations.

muthukumaranR committed:
Refactor TriageComponent for improved readability and enhanced debug logging

- Reformatted field definitions in TriageAgentOutputSchema for better alignment and clarity.
- Enhanced debug logging in the process method of TriageComponent to provide previews of input and output, improving traceability during triage operations.
- Updated import statements for consistency and clarity.

muthukumaranR committed:
Refactor prompts in prompts.py for clarity and improved structure

- Updated CLARIFYING_AGENT_PROMPT to focus on eliciting high-signal clarifications for deep literature searches.
- Revised RESEARCH_INSTRUCTION_AGENT_PROMPT to emphasize specificity and detail in research instructions.
- Streamlined TRIAGE_AGENT_PROMPT to clarify decision-making criteria for research queries.
- Enhanced CONTENT_CONDENSATION_PROMPT to ensure relevant content extraction aligns with research questions.
- Improved formatting and structure across all prompts for better readability and usability.

muthukumaranR committed:
Update ISSN whitelist file path in SourceValidatorConfig and _load_whitelist method

- Changed the default path for the ISSN whitelist JSON file from "issn_whitelist.json" to "issn_whitelist_full.json" in both the SourceValidatorConfig class and the _load_whitelist method.
- This update ensures the correct file is referenced for source validation operations.

muthukumaranR committed:
Add full ISSN whitelist JSON file for source validation

- Introduced a new file `issn_whitelist_full.json` containing a comprehensive ISSN-based journal whitelist for source validation.
- The file includes metadata and categorized ISSNs to enhance the source validation process.
- This addition supports improved accuracy in identifying valid journals for research purposes.

muthukumaranR committed:
Update deep_search_testing notebook with execution counts and output enhancements

- Set execution counts for code cells to track execution order.
- Added output messages to indicate successful setup and results found during the search process.
- Improved the structure of output streams for better readability and clarity in results presentation.

NISH1001 committed:
Bugfix full content scraping logic in deeplitsearch agent

- Remove early return after link assessment

NISH1001 committed:
Bugfix debug mode flag in scrapertool base

- Also prioritize debug param flag first then use config debug flag

muthukumaranR committed:
refactor out all search tool related code from deep search agent code.

muthukumaranR committed:
use SearchResultItem for better generalization of Search result artefacts across AKD

muthukumaranR committed:
Allow for title OR URL match for deduplication of articles instead of AND logic

muthukumaranR committed:
use LitSearchAgentOutputSchema and extra fields to add outputs of deep search agent instead of DeepLitSearchAgentOutputSchema for generalizability of search agents

muthukumaranR committed:
add extra field to LitSearchAgentOutputSchema

muthukumaranR committed:
add the research output fields to `extra` field of LitSearchAgentOutputSchema

muthukumaranR committed:
refactor managed_results to results

muthukumaranR committed:
refactor the prompt

muthukumaranR committed:
remove old tests that tested the search_tool dependencies instead of assuming pipeline usage

muthukumaranR committed:
example code to run akd/agents/search/deep_search.py

muthukumaranR committed:
change test to reflect the changes in the deep_search agent output.

muthukumaranR committed:
modify mapper tests to reflect changes on deep_search agent output schema

muthukumaranR committed:
refactor test for cleaner output while testing

NISH1001 commented on akd/_base.py:
```diff
@@ -151,7 +151,7 @@ def __init__(
         self._kwargs = kwargs
         self._post_init()
-        self.debug = getattr(config, "debug", False) or debug
+        self.debug = debug or getattr(config, "debug", False)
```
Okay. Re-prioritzing incoming debug param from consturctor -> and then to config?

muthukumaranR replied:
I think so IFF we are allowing `debug=True` from the constructor - it should be prioritized over whatever value is set the config. I see the pattern in a few other agents too where the debug is passed from the constructor and the status quo would override it i think.

NISH1001 replied:
Eh okay. That makes sense.

NISH1001 commented on akd/agents/search/components/clarification.py:
```diff
@@ -7,23 +7,29 @@
+
     query: str = Field(..., description="Query that needs clarification")
-    search_results: Optional[List[Dict]] = Field(default=None, description="Existing search results for context")
+    search_results: Optional[List[Dict]] = Field(
```
Does this be `list[searchresultitem]`??

muthukumaranR replied:
It doesn't have to be since it really only serves to provide context to the instructor - and the simpler the structure, the better / fewer tokens - but I have updated it to be `Optional[List[SearchResultItem]]` to be consistent

NISH1001 replied:
Oh interesting. Does it lead to fewer tokens? Under the hood it will convert to dict no? Or maybe we can use `set(...)` format to restrict the keys to instructor in future iteration?

NISH1001 commented on akd/agents/search/components/content_condensation.py:
```diff
@@ -0,0 +1,216 @@
+        results_with_content = [
+            r
+            for r in params.search_results
+            if r.content and len(r.content.strip()) >= 100
```
Can we move the `100` to config itself which gives more control on this filter?

muthukumaranR replied:
fixed

NISH1001 commented on akd/agents/search/components/content_condensation.py:
```diff
@@ -0,0 +1,216 @@
+                condensed_content == "[NO RELEVANT CONTENT]"
+                or len(condensed_content) < 10
+            ):
+                condensed_content = ""
```
Maybe default to original content intead of nullifying it if condensation can't be done.

`condensed_content = result.item or ""`?

muthukumaranR replied:
yes agreed. fixed.

NISH1001 commented on akd/agents/search/deep_search.py:
```diff
@@ -181,45 +115,20 @@ def __init__(
+        # Initialize search pipeline with default tools
+        if search_pipeline is None:
+            # SearchPipeline will use its default search tool internally
+            from akd.tools.search.searxng_search import SearxNGSearchTool
```
Need to import at top-level import, not within if condition. (probably claude artifact)

NISH1001 commented on akd/agents/search/deep_search.py:
```diff
@@ -570,8 +419,18 @@ async def _evaluate_research_quality(
+            )
+
         # Calculate quality score from rubrics
         from akd.agents.relevancy import (
```
need to import at top-level

muthukumaranR replied:
fixed

NISH1001 commented on akd/agents/search/deep_search.py:
```diff
@@ -686,33 +566,19 @@ async def _arun(
-            )
-
         return LitSearchAgentOutputSchema(
             results=results_as_dicts,
```
we should probably do `results` as `list[searchresultitem]` itself as well? instead of dict

NISH1001 reviewed (changes requested):
Review Pass I

muthukumaranR committed:
use searchresultitem structure

muthukumaranR committed:
pass the searchresultitem to the instructor agent

muthukumaranR committed:
fall back to original content if condensation fails

muthukumaranR committed:
Add minimum content length configuration for content condensation

muthukumaranR committed:
Add minimum content length field to ContentCondensationConfig

muthukumaranR committed:
Refactor DeepLitSearchAgent to streamline imports

muthukumaranR committed:
pass search results to clarification component

muthukumaranR committed:
Refactor DeepLitSearchAgent to return research output directly with SearchResultItem objects

muthukumaranR committed:
Update SearchAgentOutputSchema to use SearchResultItem

muthukumaranR committed:
Update tests to utilize SearchResultItem in WaterfallMapper

muthukumaranR committed:
Refactor test_deep_search.py to replace dictionary access with attribute access for SearchResultItem and update clarification component calls

muthukumaranR commented:
Interesting. The pytest passed - and somehow the CICD failed. Not sure why. Will take a deeper look.

NISH1001 committed:
Bugfix cicd integration github action

- Fix `integration-branch-sync.yml`
- Handle case where the summary line deosn't have any failed cases

NISH1001 commented on akd/agents/search/deep_search.py:
```diff
@@ -410,147 +353,59 @@ async def _execute_searches(
         queries: List[str],
         original_query: str | None = None,
         is_reformulated: bool = False,
```
how do we use `is_reformulated`? Should we add to searchpipeline for this? Curently search pipeline doens't utilize this "is_reformulated" which I presume is needed for Link assessor. Previously, within deep search agent, `reformulated_query` was passed to Link assessor input schema. 

Now in our new refactor, this is set to `None` since there's no param for it in the pipeline input.

NISH1001 commented on akd/agents/search/deep_search.py:
```diff
@@ -410,147 +353,59 @@ async def _execute_searches(
-            ss_input = SemanticScholarSearchToolInputSchema(
+        # Primary search tool
+        try:
+            tool_input = self.search_tool.input_schema(
```
maybe we can extra params like `is_reformulated`, `reformulated_query` etc as `kwargs` for time being?

NISH1001 replied:
Update: In the search pipeline, now we access these params via `kwargs`:
[see](https://github.com/NASA-IMPACT/accelerated-discovery/blob/enhance/search-pipeline/akd/tools/search/pipeline.py#L569)

Will that work? We need to send these from within the agent to the pipeline.

NISH1001 replied:
Okay i have passed this as kwargs to search_tool and search pipeline in my new branch automatically accepts that now.

NISH1001 reviewed (commented):
Comment II

NISH1001 commented on akd/agents/search/deep_search.py:
```diff
@@ -157,19 +100,17 @@ class DeepLitSearchAgent(LitBaseAgent):
         config: DeepLitSearchAgentConfig | None = None,
-        search_tool=None,
-        semantic_scholar_tool: SemanticScholarSearchTool | None = None,
+        search_pipeline: SearchPipeline | None = None,
```
Should we change this to a loose type ?
```
search_tool: SearchTool | None = None,
```

NISH1001 replied:
Added the SEarchTool union type here.

NISH1001 committed:
Change type of search_tool to SearchTool or SearchPipeline for
DeepLitSearchAgent

NISH1001 committed:
Bugfix execute search method in deep search

- Pass domain_context, reformulated_query, original_query to the tool
  execution

- Bugfix tests for new search_tool

NISH1001 commented:
@muthukumaranR i have apprved this PR. seems like there's integrati nbracnh auto merge conflict. Should i pull this branc hthere and resolve manually?
